src/tree/treemath.rs

```diff
@@ -69,40 +72,73 @@ pub(crate) fn right(index: NodeIndex, size: LeafIndex) -> Result<NodeIndex, Tree
     Ok(NodeIndex::from(r))
 }
 
+// The parent here might be beyond the right edge of the tree.
 pub(crate) fn parent_step(x: usize) -> usize {
     let k = level(NodeIndex::from(x));
     let b = (x >> (k + 1)) & 0x01;
     (x | (1 << k)) ^ (b << (k + 1))
 }
 
 pub(crate) fn parent(index: NodeIndex, size: LeafIndex) -> Result<NodeIndex, TreeMathError> {
+    node_in_tree(index, size)?;
+    unsafe_parent(index, size)
+}
+
+// This function is only safe to use if index <= size.
+// If this is not checked before calling the function, `parent` should be used.
+fn unsafe_parent(index: NodeIndex, size: LeafIndex) -> Result<NodeIndex, TreeMathError> {
     let x = index.as_usize();
     let n = size.as_usize();
     if index == root(size) {
         return Err(TreeMathError::RootHasNoParent);
     }
     let mut p = parent_step(x);
-    while p >= node_width(n) {
-        p = parent_step(p)
+    let node_width = node_width(n);
+    while p >= node_width {
+        let new_p = parent_step(p);
+        if new_p == p {
```

Comment:
Can `parent_step` of an index be the same as the index itself? I thought we should be checking here if the output is outside of the tree.

Reply:
Well, this is a purely opportunistic check. If this ends up being true, we'd end up in an endless loop. It's not an actual correctness check.